### DISCUSSION TITLE
gradient.check: how does it work?

### DISCUSSION THREAD

**Discussion by gerverska** [JenniNiku/gllvm#176] [Q&A] [answered]
Hi folks!

It's mostly in the title, but I guess I was just trying to consult the documentation about what's going on when `gradient.check = T`. 

As a follow-up question, is it inherently problematic to have a `converged with large gradients (>0.05)` warning? I'm currently in the process of fitting smaller and smaller `reltol`, and I keep getting the warning. It seems possible that some datasets could just have large gradients along LVs or covariates, right?

Here are the model details:
```
model <- gllvm(Y,
               family = 'tweedie',
               Power = NULL,
               num.lv = 2,
               offset = offset,
               gradient.check = T,
               control = list(reltol = 1e-25,
                              max.iter = 12000),
               control.start = list(n.init = 3))
```

**Comment by BertvanderVeen** [comment, marked as answer]
`gradient.check` uses the gradient, I.e. The vector of first derivatives to the likelihood, to check if the model has converged. At the maximum the surface is flat, so the gradient should be zero.

If the gradient is not zero (by some tolerance) for some parameters, the model has not properly converged. This can happen due to a variety of reasons, for example, due to a poor model fit in one way or another. 

With many parameters it tends to be difficult to get the gradient for all parameters below some tolerance, and whether it really is problematic or not (in my experience) depends on the (number and type) of parameters that this applies to. 

You can check with parameters are not playing nice by checking model$TMBfn$gr() while using the names from the model$TMBfn$par object. 

Reltol values lower than machine accuracy as indicated in your post will not achieve anything.

**Reply by gerverska** [reply]
Gotcha--I didn't even think of double-precision limitations, thanks.

Here's a plot for each parameter:
```
gradient.check <- data.frame(gradient = model$TMBfn$gr() %>% .[1, ],
                             par = model$TMBfn$par |> names())

gradient.check |> 
ggplot(aes(x = par, y = gradient)) +
    geom_jitter(alpha = 0.25,
                width = 0.1) +
    scale_y_continuous(n.breaks = 8) +
    xlab("\nParameter") +
    ylab("Gradient\n") +
    theme_classic() +
    base.theme
```
![image](https://github.com/JenniNiku/gllvm/assets/29261367/b1ece5f3-88dd-45dd-a99f-e26de5f1e3de)

I'm not too sure how to interpret this, but it seems that values for `Au`, `b`, `lambda`, `lg_phi`, and `u` converge on 0, while `ePower` and `sigmaLV` only seem to have single values (because I assume their estimates aren't affected by extra iterations?) with gradients that exceed the absolute value of 0.05.

Or do values for `Au`, `b`, `lambda`, `lg_phi`, and `u` need to plotted by row index to really confirm convergence?

**Reply by gerverska** [reply]
By the way, I'm using the CRAN version here, but I can swap to the current GH commit.

**Reply by BertvanderVeen** [reply]
That looks fine. All are sufficiently close to zero, so I would not take the warning too seriously.

**Reply by gerverska** [reply]
Excellent, this is good to hear. What do you think is triggering the gradient warning then? Are the `ePower` and `sigmaLV` terms capable of triggering the warning, or are the couple of stray gradient values from `lambda` and `u` the culprits? I guess I'm assuming that absolute values > 0.05 trigger the warning--is this actually the case, or are only positive values considered?

**Reply by BertvanderVeen** [reply]
Yes, absolute values.

**Reply by gerverska** [reply]
Gotcha, and what affects which parameters are included with `model$TMBfn$gr()` and `model$TMBfn$par`? Below is the figure whenever I changed the method to `LA`, and it seems some of the terms from before are no longer included.

![image](https://github.com/JenniNiku/gllvm/assets/29261367/283e54a8-3915-4556-9c0d-e7bd85b8ab15)

**Reply by BertvanderVeen** [reply]
The VA method includes additional parameters for the approximation; in LA something similar happens but they are not (explicitly) available, and do not show up in the gradient.

**Reply by gerverska** [reply]
Thanks Bert!
